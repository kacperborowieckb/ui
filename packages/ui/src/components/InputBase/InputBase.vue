<template>
  <input
    v-model="model"
    class="placeholder:text-muted-foreground py-1/2 h-full min-w-[120px] flex-1 px-2 focus:outline-0"
    type="text"
    :placeholder="placeholder"
    :disabled="disabled"
  >
</template>

<script lang="ts">
export interface InputBaseProps {
  placeholder?: string
  disabled?: boolean
}
</script>

<script setup lang="ts">
defineProps<InputBaseProps>()

const model = defineModel<string>()
</script>
