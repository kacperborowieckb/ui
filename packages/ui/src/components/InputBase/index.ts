export {
  default as InputBase,
  type InputBaseProps,
} from './InputBase.vue'
