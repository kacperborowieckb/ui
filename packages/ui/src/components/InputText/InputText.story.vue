<template>
  <Story>
    <Variant title="Default">
      <div class="p-4">
        <InputText
          v-model="username"
          class="max-w-[324px]"
          placeholder="Pick your username"
          label="Username"
          errorMessage="Invalid username"
          :validator="/^[a-zA-Z0-9]*$/"
        />
      </div>
    </Variant>
    <Variant title="Error state">
      <div class="p-4">
        <InputText
          hasError
          class="max-w-[324px]"
          placeholder="Pick your username"
          label="Username"
          errorMessage="Invalid username"
        />
      </div>
    </Variant>
    <Variant title="Disabled">
      <div class="p-4">
        <InputText
          disabled
          class="max-w-[324px]"
          placeholder="Pick your username"
          label="Username"
        />
      </div>
    </Variant>
  </Story>
</template>

<script setup lang="ts">
import { ref } from 'vue'

import InputText from './InputText.vue'

const username = ref('')
</script>
