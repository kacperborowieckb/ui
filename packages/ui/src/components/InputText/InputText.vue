<template>
  <InputWrapper
    :label="label"
    :disabled="disabled"
    :hasError="isInvalid"
    :errorMessage="errorMessage"
  >
    <InputBase
      v-model="inputValue"
      :placeholder="placeholder"
      :disabled="disabled"
    />
  </InputWrapper>
</template>

<script lang="ts">
import { computed } from 'vue'

import { InputBase } from '../InputBase'
import { InputWrapper } from '../InputWrapper'

export interface InputTextProps {
  placeholder?: string
  label?: string
  disabled?: boolean
  validator?: RegExp | ((chipValue: string) => boolean)
  hasError?: boolean
  errorMessage?: string
}
</script>

<script setup lang="ts">
const props = defineProps<InputTextProps>()

const inputValue = defineModel<string>({ default: '' })
const error = defineModel<boolean>('error', { default: false })

const isInvalid = computed(() => {
  return props.hasError || !validateInputValue(inputValue.value)
})

function validateInputValue(value: string) {
  if (!props.validator)
    return true

  if (props.validator instanceof RegExp) {
    error.value = !props.validator.test(value)
  }

  if (props.validator instanceof Function) {
    error.value = !props.validator(value)
  }

  return !error.value
}
</script>
