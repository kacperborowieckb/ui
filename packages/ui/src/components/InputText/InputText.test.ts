import { shallowMount } from '@vue/test-utils'

import { InputBase } from '../InputBase'
import InputText from './InputText.vue'
import type { InputTextProps } from './InputText.vue'

describe('inputText', () => {
  function getWrapper(props?: InputTextProps) {
    return shallowMount(InputText, {
      props,
      global: {
        stubs: {
          InputWrapper: {
            template: '<input-wrapper-stub><slot /></input-wrapper-stub>',
          },
        },
      },
    })
  }

  it('should render correctly', () => {
    const wrapper = getWrapper({
      label: 'mockLabel',
      placeholder: 'mockPlaceholder',
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly in disabled state', () => {
    const wrapper = getWrapper({
      label: 'mockLabel',
      placeholder: 'mockPlaceholder',
      disabled: true,
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly in error state', () => {
    const wrapper = getWrapper({
      label: 'mockLabel',
      placeholder: 'mockPlaceholder',
      hasError: true,
      errorMessage: 'mockErrorMessage',
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should correctly define value model', async () => {
    const wrapper = getWrapper()

    await wrapper.findComponent(InputBase).setValue('mockValue')

    expect(wrapper.emitted('update:modelValue')).toStrictEqual([['mockValue']])
  })

  it('should correctly define error model respecting the validator function', async () => {
    const wrapper = getWrapper({
      validator: val => val.length > 4,
    })

    await wrapper.findComponent(InputBase).setValue('moc')

    expect(wrapper.emitted('update:error')).toStrictEqual([[true]])
  })

  it('should correctly define error model respecting the validator regexp', async () => {
    const wrapper = getWrapper({
      validator: /^\d*$/,
    })

    await wrapper.findComponent(InputBase).setValue('24d')

    expect(wrapper.emitted('update:error')).toStrictEqual([[true]])
  })
})
