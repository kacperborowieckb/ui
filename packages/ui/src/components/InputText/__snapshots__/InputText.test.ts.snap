// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`inputText > should render correctly 1`] = `
"<input-wrapper-stub label="mockLabel" disabled="false" haserror="false">
  <input-base-stub modelvalue="" placeholder="mockPlaceholder" disabled="false"></input-base-stub>
</input-wrapper-stub>"
`;

exports[`inputText > should render correctly in disabled state 1`] = `
"<input-wrapper-stub label="mockLabel" disabled="true" haserror="false">
  <input-base-stub modelvalue="" placeholder="mockPlaceholder" disabled="true"></input-base-stub>
</input-wrapper-stub>"
`;

exports[`inputText > should render correctly in error state 1`] = `
"<input-wrapper-stub label="mockLabel" disabled="false" haserror="true" errormessage="mockErrorMessage">
  <input-base-stub modelvalue="" placeholder="mockPlaceholder" disabled="false"></input-base-stub>
</input-wrapper-stub>"
`;
